### PR TITLE
[FEATURE] Additional Rule-Based Profiler Parameter/Variable Access Methods

### PR DESCRIPTION
**Pull request by alexsherstinsky**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- GREAT-735/GREAT-791
-
-

After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/rule_based_profiler/types/domain.py [hunk 1]
```diff
@@ -121,11 +121,23 @@ def id(self) -> str:
         return IDDict(self.to_json_dict()).to_id()
 
     def to_json_dict(self) -> dict:
+        details: dict = {}
+
+        key: str
+        semantic_type: Union[str, SemanticDomainTypes]
+        for key, semantic_type in self["details"].items():
+            if isinstance(semantic_type, str):
```
Comment (anthonyburdi) [inline great_expectations/rule_based_profiler/types/domain.py]: Does this work if the string from details is for something other than a SemanticDomainType? Would this code then try to convert it to a SemanticDomainType?

Comment (anthonyburdi) [reply great_expectations/rule_based_profiler/types/domain.py]: It would be great to have a test specifically for this method with and without a semantic_type to be sure.

File: tests/rule_based_profiler/parameter_builder/test_parameter_computations.py [hunk 2]
```diff
@@ -150,108 +148,3 @@ def test_bootstrap_point_estimate_efficacy(
             atol=EFFICACY_TOLERANCE,
             err_msg=f"Actual value of {actual_false_positive_rates[distribution]} differs from expected value of {false_positive_rate} by more than {ATOL + EFFICACY_TOLERANCE * abs(actual_false_positive_rates[distribution])} tolerance.",
         )
-
-
-def test_bootstrap_point_estimate_bias_corrected_efficacy(
```
Comment (anthonyburdi) [inline tests/rule_based_profiler/parameter_builder/test_parameter_computations.py]: Is this test no longer relevant / necessary?

**Review by anthonyburdi — COMMENTED** [review]
Largely this looks great and I love the use of ParameterBuilderConfig instead of dict. Some questions that might require changes so just leaving comments.

File: great_expectations/rule_based_profiler/types/domain.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/types/domain.py]: @anthonyburdi It does work -- either `str` or `SemanticType`.  One of my new tests were failing, so this fix (which was a long-standing typo) made them work.  I will add a test (in the next commit).  Thanks.

File: tests/rule_based_profiler/parameter_builder/test_parameter_computations.py (hunk 2 above)
Comment (alexsherstinsky) [reply tests/rule_based_profiler/parameter_builder/test_parameter_computations.py]: @anthonyburdi @NathanFarmer I found that all 3 (three) tests used methods that no longer existed anywhere!  One of them could be resurrected by removing the underscore from the method's name (which is what I did).  @NathanFarmer could you please shed some light on this?  I could not even run `pytest` with those broken imports.  Thanks.

Comment (NathanFarmer) [reply tests/rule_based_profiler/parameter_builder/test_parameter_computations.py]: I removed the broken/unnecessary pieces in PR #4818. But why didn't CI/CD catch this in PR #4742? I wonder if we are having problems with `dgtest`.

File: great_expectations/rule_based_profiler/types/domain.py (hunk 1 above)
Comment (anthonyburdi) [reply great_expectations/rule_based_profiler/types/domain.py]: @alexsherstinsky That makes sense if the str is one of the named semantic types. But what if the value in "details" is not one of the SemanticDomainTypes?

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/types/domain.py]: @anthonyburdi The test I added covers this situation.  Thanks.

File: tests/rule_based_profiler/parameter_builder/test_parameter_computations.py (hunk 2 above)
Comment (alexsherstinsky) [reply tests/rule_based_profiler/parameter_builder/test_parameter_computations.py]: Thanks, @NathanFarmer -- the merge was smooth.

File: tests/rule_based_profiler/domain_builder/test_cardinality_checker.py [hunk 3]
```diff
@@ -83,7 +83,8 @@ def test_cardinality_checker_instantiation_valid_max_proportion_unique_parameter
 
 def test_cardinality_checker_instantiation_invalid_limit_mode_parameter():
     with pytest.raises(ProfilerConfigurationError) as excinfo:
-        _: CardinalityChecker = CardinalityChecker(
+        # noinspection PyUnusedLocal,PyTypeChecker
+        cardinality_checker: CardinalityChecker = CardinalityChecker(
             limit_mode="&^*%not_a_valid_mode^&*("
         )
```
Comment (anthonyburdi) [inline tests/rule_based_profiler/domain_builder/test_cardinality_checker.py:89]: Since we are not doing assignment here (and in all the other examples) maybe it's cleaner to just remove the assignment altogether e.g.:
```suggestion
        CardinalityChecker(limit_mode="&^*%not_a_valid_mode^&*(")
```

Comment (alexsherstinsky) [reply tests/rule_based_profiler/domain_builder/test_cardinality_checker.py:89]: @anthonyburdi I find it nice to indicate that "the thing" returns something.  So the choice for me was to use `_ =` without a type hint or a dummy return value with a type hint -- I opted for the latter in the name of interpretability.  If you feel strongly in favor of a different way, please let me know, and I will change it.  Thanks!

File: tests/conftest.py [hunk 4]
```diff
@@ -3662,7 +3660,6 @@ def bobby_columnar_table_multi_batch(empty_data_context):
         ExpectationConfiguration(
             **{
                 "expectation_type": "expect_column_values_to_match_regex",
-                "meta": {"notes": {"format": "markdown", "content": None}},
```
Comment (NathanFarmer) [inline tests/conftest.py:3665]: Why was this necessary? Why didn't our tests fail before you did this?

Comment (NathanFarmer) [reply tests/conftest.py:3665]: Most likely we have only been testing pieces of `ExpectationConfiguration`, but shouldn't we address this in the test instead of the fixture?

**Review by NathanFarmer — COMMENTED** [review]
Looks good! Only one question from me, and then leaving the rest to @anthonyburdi's feedback.

File: tests/conftest.py (hunk 4 above)
Comment (alexsherstinsky) [reply tests/conftest.py:3665]: @NathanFarmer I saw a `PyCharm` error -- duplicated key; so I deleted it.  Our tests did not fail, because the dictionary does not forbid specifying the same key multiple times -- the last one is used.  This is one of those typos that an IDE catches for us.  Thanks.

File: great_expectations/rule_based_profiler/types/domain.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/types/domain.py]: @anthonyburdi I updated the functionality as we discussed and added more tests.  Thanks.

File: great_expectations/rule_based_profiler/types/domain.py [hunk 5]
```diff
@@ -124,13 +126,17 @@ def to_json_dict(self) -> dict:
...
+        for key, value in self["details"].items():
+            if key == INFERRED_SEMANTIC_TYPE_KEY:
+                semantic_type: Union[str, SemanticDomainTypes]
+                if isinstance(value, str):
+                    semantic_type = value.lower()
+                    semantic_type = SemanticDomainTypes(semantic_type)
+                else:
+                    semantic_type = value
+
+                details[key] = semantic_type.value
```
Comment (anthonyburdi) [inline great_expectations/rule_based_profiler/types/domain.py:139]: Should we also do the else clause so we don't lose non semantic type details?:
```suggestion
        value: Any
        for key, value in self["details"].items():
            if key == INFERRED_SEMANTIC_TYPE_KEY:
                semantic_type: Union[str, SemanticDomainTypes]
                if isinstance(value, str):
                    semantic_type = value.lower()
                    semantic_type = SemanticDomainTypes(semantic_type)
                else:
                    semantic_type = value

                details[key] = semantic_type.value
             else:
                 details[key] = value
```

File: great_expectations/rule_based_profiler/types/domain.py [hunk 6]
```diff
@@ -12,6 +12,8 @@
     filter_properties_dict,
 )
 
+INFERRED_SEMANTIC_TYPE_KEY: str = "inferred_semantic_domain_type"
```
Comment (anthonyburdi) [inline great_expectations/rule_based_profiler/types/domain.py:15]: 👍

File: tests/rule_based_profiler/domain_builder/test_domain_builder.py [hunk 7]
```diff
@@ -276,6 +276,7 @@ def test_column_domain_builder_with_simple_semantic_type_included(
         data_context=data_context,
     )
     domains: List[Domain] = domain_builder.get_domains(variables=variables)
+    print(f"\n[ALEX_TEST] [WOUTPUT] WOUTPUT:\n{domains} ; TYPE: {str(type(domains))}")
```
Comment (anthonyburdi) [inline tests/rule_based_profiler/domain_builder/test_domain_builder.py]:
```suggestion
```

File: tests/rule_based_profiler/domain_builder/test_domain_builder.py [hunk 8]
```diff
@@ -359,6 +360,60 @@ def test_semantic_domain_comparisons():
         in str(excinfo.value)
     )
 
+    domain_e: Domain = Domain(
```
Comment (anthonyburdi) [inline tests/rule_based_profiler/domain_builder/test_domain_builder.py]: I think I would break these out into a separate `test_domain_json_serialization` or similar and also test the contents of the json returned

File: great_expectations/rule_based_profiler/types/domain.py (hunk 5 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/types/domain.py:139]: @anthonyburdi great catch -- in next commit -- thanks!

File: tests/rule_based_profiler/domain_builder/test_domain_builder.py (hunk 8 above)
Comment (alexsherstinsky) [reply tests/rule_based_profiler/domain_builder/test_domain_builder.py]: @anthonyburdi Done!

**Review by anthonyburdi — APPROVED** [review]
LGTM! Thanks for the changes
